cpp/src/quantiles/quantile.cu

```diff
@@ -0,0 +1,109 @@
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <quantiles/quantiles_util.hpp>
+#include "cudf/utilities/error.hpp"
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+using ScalarResult = double;
+
+struct quantile_functor
+{
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
+    operator()(Args&&... args)
+    {
+        CUDF_FAIL("Only numeric types are supported in quantiles.");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
+    operator()(column_view const& input,
+               double percent,
+               interpolation interp,
+               order_info column_order,
+               rmm::mr::device_memory_resource *mr =
+                 rmm::mr::get_default_resource(),
+               cudaStream_t stream = 0)
+    {
+        if (input.size() == input.null_count()) {
+            return std::make_unique<numeric_scalar<ScalarResult>>(0, false, stream);
+        }
+
+        if (input.size() == 1) {
+            auto result = get_array_value<ScalarResult>(input.begin<T>(), 0);
+            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
+        }
+
+        auto valid_count = input.size() - input.null_count();
+
+        if (column_order.is_sorted != sorted::YES)
+        {
+            table_view const in_table { { input } };
+            auto sortmap = sorted_order(in_table, { order::ASCENDING }, { null_order::AFTER });
+            auto sortmap_begin = sortmap->view().begin<size_type>();
+            auto input_begin = input.begin<T>();
+
+            auto selector = [&](size_type location) {
+                auto idx = detail::get_array_value<size_type>(sortmap_begin, location);
+                return detail::get_array_value<T>(input_begin, idx);
+            };
+
+            auto result = select_quantile<ScalarResult>(selector, valid_count, percent, interp);
+            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
+        }
+
+        auto input_begin = column_order.ordering == order::ASCENDING
+            ? input.begin<T>() + (column_order.null_ordering == null_order::BEFORE ? input.null_count() : 0)
+            : input.begin<T>() - (column_order.null_ordering == null_order::AFTER  ? input.null_count() : 0) + input.size() - 1;
+
+        auto selector = [&](size_type location) {
+            return get_array_value<T>(input_begin, column_order.ordering == order::ASCENDING ? location : -location);
+        };
+
+        auto result = select_quantile<ScalarResult>(selector, valid_count, percent, interp);
+        return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
+    }
+};
+
+} // anonymous namespace
+} // namespace detail
+
+std::unique_ptr<scalar>
+quantile(column_view const& input,
```

Comment:
So why do we have `quantiles` for tables, and only `quantile` for single columns? Why not get rid of the single column version? (after making the change to allow arithmetic interpolation when all columns are arithmetic)

Reply:
I thought the same thing, and have started working on consolidating the logic. I wanted to get a working version of what was described in #4011 before continuing with that refactor, in case it’s needed to unblock someone else.

Reply:
See https://github.com/rapidsai/cudf/issues/4011#issuecomment-582991395